lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -882,158 +878,171 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
             Cardano.TxReturnCollateral _ _ ->
                throwE ErrBalanceTxExistingReturnCollateral
 
-    -- | Select assets to cover the specified balance and fee.
-    --
-    -- If the transaction contains redeemers, the function will also ensure the
-    -- selection covers the fees for the maximum allowed execution units of a
-    -- transaction. For this, and other reasons, the selection may include too
-    -- much ada.
-    selectAssets'
-        :: Cardano.AnyCardanoEra
-        -> [W.TxOut]
-        -> UTxOSelection WalletUTxO
-        -- ^ Describes which utxos are pre-selected, and which can be used as
-        -- inputs or collateral.
-        -> Cardano.Value -- Balance to cover
-        -> Cardano.Lovelace -- Current minfee (before selecting assets)
-        -> StdGenSeed
-        -> Either (SelectionError WalletSelectionContext) Selection
-    selectAssets' era outs utxoSelection balance fee0 seed =
-        flip evalRand (stdGenFromSeed seed)
-            $ runExceptT
+    fromCardanoLovelace (Cardano.Lovelace l) = Coin.unsafeFromIntegral l
+
+-- | Select assets to cover the specified balance and fee.
+--
+-- If the transaction contains redeemers, the function will also ensure the
+-- selection covers the fees for the maximum allowed execution units of a
+-- transaction. For this, and other reasons, the selection may include too
+-- much ada.
+selectAssets
+    :: forall era changeState
+     . Cardano.IsCardanoEra era
+    => RecentEra era
+    -> ProtocolParameters era
+    -> TxWitnessTag
+    -> [W.TxOut]
+    -> [Redeemer]
+    -> UTxOSelection WalletUTxO
+    -- ^ Describes which utxos are pre-selected, and which can be used as
+    -- inputs or collateral.
+    -> Cardano.Value -- Balance to cover
+    -> W.Coin -- Current minfee (before selecting assets)
```

Comment:
So at _some point_ we should presumably switch to ledger types for these

Reply:
> So at some point we should presumably switch to ledger types for these

Definitely!

It would **_also_** be nice to avoid reconverting back to `W.Coin` and `W.TokenBundle` before calling into the coin selection algorithm. 😄 

My current inclination is to steadily remove the coin selection algorithm's dependence on these concrete types in favour of minimal class constraints. Then you could supply any types, without any conversions, so long as the constraints are satisfied.